### ISSUE TITLE
[Bug]: Crash in Combobox when allowFreeform true and options are changed

### ISSUE THREAD

**Issue by rcbevans** [microsoft/fluentui#25932]
### Library

React / v8 (@fluentui/react)

### System Info

```shell
Commit 3c754611eccb854647820ce5b5fd8b5f6b0cf1c0 introduced a bug causing a crash in the ComboBox component when allowFreeform is set to true, and the user of the Combobox is asynchronously filtering/populating the ComboBox options using the user input.


Uncaught TypeError: Cannot read properties of undefined (reading 'id')
    at ComboBoxInternal._getAriaActiveDescendantValue (ComboBox.tsx:2368:1)
    at ComboBoxInternal._this._renderComboBoxWrapper (ComboBox.tsx:624:1)
    at ComboBoxInternal.render
```

I have added on comment on the MR that introduced the crash: https://github.com/microsoft/fluentui/commit/3c754611eccb854647820ce5b5fd8b5f6b0cf1c0#r92388055

The line to blame is

```
descendantText = options[currentPendingValueValidIndex].id ?? this._id + '-list' + currentPendingValueValidIndex;

`currentPendingValueValidIndex` is checked to see if it is -1, but not checked to ensure it is within the bounds of the options array.  As a consequence, if the options are changed after `_processInputChangeWithFreeform` has executed, such that `currentPendingValueValidIndex` is greater than the length of the options array, the component will crash.
```

### Are you reporting Accessibility issue?

None

### Reproduction

https://codesandbox.io/s/react-fiddle-forked-yujk10?file=/src/index.tsx

## Actual Behavior
If options are changed based on the ComboBox text, ComboBox crashes.

## Expected Behavior
ComboBox shouldn't crash

### Logs

```shell
TypeError
options[currentPendingValueValidIndex] is undefined
$csb$eval/ComboBoxInternal.prototype._getAriaActiveDescendantValue@https://yujk10.csb.app/node_modules/
fluentui/react/lib-commonjs/components/ComboBox/ComboBox.js:1689:13
ComboBoxInternal/_this._renderComboBoxWrapper@https://yujk10.csb.app/node_modules/
fluentui/react/lib-commonjs/components/ComboBox/ComboBox.js:202:877
$csb$eval/ComboBoxInternal.prototype.render@https://yujk10.csb.app/node_modules/
fluentui/react/lib-commonjs/components/ComboBox/ComboBox.js:942:36
finishClassComponent
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:17485:31
updateClassComponent
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:17435:44
beginWork
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:19073:16
callCallback
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:3945:14
invokeGuardedCallbackDev
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:3994:16
invokeGuardedCallback
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:4056:31
beginWork$1
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:23964:28
performUnitOfWork
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:22776:12
workLoopSync
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:22707:22
renderRootSync
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:22670:7
performSyncWorkOnRoot
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:22293:18
flushSyncCallbackQueueImpl/<
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:11327:26
unstable_runWithPriority
https://yujk10.csb.app/node_modules/scheduler/cjs/scheduler.development.js:468:12
runWithPriority$1
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:11276:10
flushSyncCallbackQueueImpl
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:11322:26
flushSyncCallbackQueue
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:11309:3
flushPassiveEffectsImpl
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:23620:3
unstable_runWithPriority
https://yujk10.csb.app/node_modules/scheduler/cjs/scheduler.development.js:468:12
runWithPriority$1
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:11276:10
flushPassiveEffects
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:23447:14
flushDiscreteUpdates
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:22356:3
finishEventHandler
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:3714:5
batchedEventUpdates
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:3748:5
dispatchEventForPluginEventSystem
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:8507:22
attemptToDispatchEvent
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:6005:36
dispatchEvent
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:5924:41
unstable_runWithPriority
https://yujk10.csb.app/node_modules/scheduler/cjs/scheduler.development.js:468:12
runWithPriority$1
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:11276:10
discreteUpdates$1
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:22413:14
discreteUpdates
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:3756:12
dispatchDiscreteEvent
https://yujk10.csb.app/node_modules/react-dom/cjs/react-dom.development.js:5889:18
```

### Requested priority

Blocking

### Products/sites affected

Meetingflow.com

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

**Comment by rcbevans** [comment]
Whilst a simple bounds check could be added to avoid the code crashing, the real issue is that `currentPendingValueValidIndex` is not being reset if the `options` array is changed.

**Comment by hongxuantang** [comment]
We're also running into this issue after updating to 8.103.10:

```
Uncaught TypeError: Cannot read properties of undefined (reading 'id')
    at ComboBoxInternal._getAriaActiveDescendantValue (ComboBox.js?04a6:1694:1)
```

**Comment by rcbevans** [comment]
I'm wiling to commit and PR for this if I can get some guidance from the FluentUI team on the "right" way to fix this.

Ideally, if the caller changes the selection options, the selected indices should be clamped/filtered to only those which are still valid, or reset entirely, if the underlying data has changed.

I see ComboBox is a legacy React Class components, so I would have previously used componentWillUpdate to check if the selection length changed and set the selected index with a clamped in-bounds value, but this lifecycle hook is deprecated as of React 17.

**Comment by thoffmann-fms** [comment]
Having the same crash.

**Comment by sopranopillow** [comment]
Hey @rcbevans, I'll look into this in this sprint. Sorry for the delay.

**Comment by WhiteAdy** [comment]
> Hey @rcbevans, I'll look into this in this sprint. Sorry for the delay.

Hello, any updates on this?

**Comment by sopranopillow** [comment]
Hey @WhiteAdy, sorry for the delay. As mentioned above, sadly a bounds check is not the fix for it. I currently have a possible fix on a draft PR #26574, but I'm talking with the team to see if we will have to fix it deeper.

**Comment by WhiteAdy** [comment]
Much appreciated! I'm using this component pretty extensively throughout different forms in the app and replacing them all would be a lot of pain.
